lib/multigrid.cpp

```diff
@@ -512,18 +510,57 @@ namespace quda
+        coarse_solver = new PreconditionedSolver(*solver, *matCoarseSmoother->Expose(), *param_coarse_solver, profile,
+                                                 coarse_prefix);
       } else {
-	Solver *solver = Solver::create(*param_coarse_solver, *matCoarseResidual, *matCoarseResidual, *matCoarseResidual, profile);
+        Solver *solver
+          = Solver::create(*param_coarse_solver, *matCoarseResidual, *matCoarseResidual, *matCoarseResidual, profile);
         sprintf(coarse_prefix, "MG level %d (%s): ", param.level + 1,
                 param.mg_global.location[param.level + 1] == QUDA_CUDA_FIELD_LOCATION ? "GPU" : "CPU");
         coarse_solver = new PreconditionedSolver(*solver, *matCoarseResidual->Expose(), *param_coarse_solver, profile, coarse_prefix);
       }
 
+      setOutputPrefix(prefix); // restore since we just popped back from coarse grid
+
+      if (param.level == param.Nlevel - 2 && param.mg_global.use_eig_solver[param.level + 1]) {
+
+        // Test if a coarse grid deflation space needs to be transferred to the coarse solver to prevent recomputation
+        int defl_size = evecs.size();
+        auto &coarse_solver_inner = *reinterpret_cast<PreconditionedSolver *>(coarse_solver)->ExposeSolver();
+        if (defl_size > 0 && transfer && param.mg_global.preserve_deflation) {
+          coarse_solver_inner.deflate_compute = false;
+        }
+
+        // Run a first dummy solve so that the deflation space is constructed and computed if needed during the MG setup
+        spinorNoise(*r_coarse, *coarse->rng, QUDA_NOISE_UNIFORM);
+        param_coarse_solver->maxiter = 1; // do a single iteration on the dummy solve
+        (*coarse_solver)(*x_coarse, *r_coarse);
+        setOutputPrefix(prefix); // restore since we just popped back from coarse grid
+        param_coarse_solver->maxiter = param.mg_global.coarse_solver_maxiter[param.level + 1];
+
+        // Transfer vectors to deflation space
+        if (defl_size > 0) {
+          if (getVerbosity() >= QUDA_VERBOSE)
+            printfQuda("Transferring deflation space size %d to coarse solver\n", defl_size);
+          // Create space in coarse solver to hold deflation space, destroy space in MG.
+          for (int i = 0; i < defl_size; i++) coarse_solver_inner.evecs.push_back(evecs[i]);
+          evecs.resize(0);
+          if (defl_size > 0 && transfer && param.mg_global.preserve_deflation) {
+            // Run a second dummy solve so that the deflation eigenvalues are recomputed
+            coarse_solver_inner.recompute_evals = true;
+            param_coarse_solver->maxiter = 1; // do a single iteration on the dummy solve
+            (*coarse_solver)(*x_coarse, *r_coarse);
+            setOutputPrefix(prefix); // restore since we just popped back from coarse grid
+            param_coarse_solver->maxiter = param.mg_global.coarse_solver_maxiter[param.level + 1];
+          }
+        }
+      }
+
```

Comment:
I find the above code (lines 528-563) rather hard to follow.  Moreover, we have two calls to the coarse solver, though I realize the deflation space is only constructed once here.  Could this be cleaned up a bit with some more comments for the reader, and perhaps only ever call the coarse solver once?

Reply:
Given my comment above about `recompute_evals`, `evecs`, etc. being made private.  Perhaps setting or removing the evectors from the inner solver should be made a method?